narwhals/_spark_like/dataframe.py

```diff
@@ -187,7 +187,7 @@ def _iter_columns(self) -> Iterator[Column]:
 
     @property
     def columns(self: Self) -> list[str]:
-        return list(self.schema)
+        return list(self.native.columns)
```

Comment:
`self.schema` gets cached, can we revert this?

that, and perhaps add a comment explaining why we use `self.schema` in case anyone changes this in the future

Reply:
There is an issue with both SQLFrame when using `self.schema`: in `SparkLikeLazyGroupBy.agg`, when getting to the last line, `.rename` fails due to:

> NotImplementedError: This engine does not support schema inference likely since it does not have an active connection.

which I didn't really bother to investigate deeper considering all the remaining issues and edge case of the PR.

Maybe we could even do:

```py
return list(self.schema) if self._cached_schema else list(self.native.columns)
```

Aside the failing case, why should we collect the schema when only for column names?

Reply:
> Maybe we could even do:

hmmm yeah that looks nice, and then also do the same caching for `.columns`? we call `.columns` a few times in `join`, probably good to cache it. but, ok for a separate pr